README.md
# zephjs-loading

An example of a loading screen for usage with the [ZephJS Framework](https://github.com/awesomeeng/zephjs).  This can be used as is or modified to suit your own tastes.

## Contents
 - [Setup](#setup)
 - [Usage](#usage)
 - [How It Works](#how-it-works)
 - [Examples](#examples)
 - [Awesome Engineering](#the-awesome-engineering-company)
 - [Support and Help](#support-and-help)
 - [License](#license)

## Setup

zephjs-loading is available as an npm package:

```shell
npm install zephjs-loading
```

Once installed find the `zephjs-loading.js` file and copy it into your local project as needed.

## Usage

Within your HTML file, include the following as close to top of the file as possible:

```html
<script src="zephjs-loading.js" type="module"></script>
```

## How It Works

ZephJS has a [set of events](https://github.com/awesomeeng/zephjs/blob/master/docs/ComponentConcepts.md#zephjs-events) which fire on the browser `document` object. This includes the `zeph:ready` event to indicate that ZephJS was loaded and any components pending definition via ZephJS have completed.

zephjs-loading taps that event as well as others in computing how much is left to load and if loading is complete.

Please see the source code of `zephjs-loading.js` for more details.

## Examples

zephjs-loading ships with a set of examples for your reference.

 - [Test](./test/index.html): The test example which is used during zephjs-loading development.

## The Awesome Engineering Company

zephjs-loading is written and maintained by The Awesome Engineering Company. We believe in building clean, configurable, creative software for engineers and architects and customers.

To learn more about The Awesome Engineering Company and our suite of products, visit us on the web at https://awesomeeng.com.

## Support and Help

## License

zephjs-loading is released under the MIT License. Please read the  [LICENSE](./LICENSE) file for details.


RELEASE_NOTES.md
# zephjs-loading Release Notes

#### **Version 1.0.0**

 - Initial release.


zephjs-loading.js
let loader = ()=>{
	// add loading layer
	let container = document.createElement("div");
	container.style = "z-index: 9999999999; position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px; background: white; font-family: sans-serif";
	document.body.appendChild(container);

	// add title
	let title = document.createElement("div");
	title.style = "font-size: 48px; font-weight: bold; color: black; padding: 25px 25px 10px 25px;letter-spacing: 8px;text-shadow: #A0A0A0 2px 2px 5px;";
	title.textContent = "loading...";
	container.appendChild(title);

	// add progress bar
	let bar = document.createElement("div");
	bar.style = "border: 1px solid #A0A0A0; min-width: 200px; max-width: 400px; width: 50%; height: 20px;margin: 0px 25px; overflow: hidden; padding: 1px 0px;";
	container.appendChild(bar);
	let inner = document.createElement("div");
	inner.style = "background: #3399FF; height: 100%; transition: width 0.5s, margin-left 0.25s;";
	inner.style.width = "50%";
	bar.appendChild(inner);

	// add footer
	let footer = document.createElement("a");
	footer.style = "position: absolute; bottom: 0px; left: 0px; right: 0px; background: black; color: white; font-size: 10px; padding: 5px 5px; text-align: right;";
	footer.textContent = "Built with ZephJS!";
	footer.setAttribute("href","http://zephjs.com");
	container.appendChild(footer);

	let loading = null;
	let loaded = null;
	let tick = 1;

	const update = ()=>{
		if (loading===null) {
			tick += 1;
			let start = (tick%150)-50;
			inner.style.width = "50%";
			inner.style.marginLeft = start+"%";
			setTimeout(update,20);
		}
		else {
			inner.style.marginLeft = "0px";
			inner.style.width = ((100/loading)*loaded)+"%";
		}
	};
	update();

	// wait for zeph:ready event
	document.addEventListener("zeph:loading",()=>{
		loading = (loading===null ? 0 : loading) + 1;
		update();
	});
	document.addEventListener("zeph:loaded",()=>{
		loaded = (loaded===null ? 0 : loaded) + 1;
		update();
	});
	document.addEventListener("zeph:ready",()=>{
		update();
		container.remove();
	});
};

let attempt = ()=>{
	if (document.body) {
		loader();
	}
	else {
		setTimeout(attempt,0);
	}
};

attempt();


index.js
let loader = ()=>{
	// add loading layer
	let container = document.createElement("div");
	container.style = "z-index: 9999999999; position: absolute; top: 0px; bottom: 0px; left: 0px; right: 0px; background: white; font-family: sans-serif";
	document.body.appendChild(container);

	// add title
	let title = document.createElement("div");
	title.style = "font-size: 48px; font-weight: bold; color: black; padding: 25px 25px 10px 25px;letter-spacing: 8px;text-shadow: #A0A0A0 2px 2px 5px;";
	title.textContent = "loading...";
	container.appendChild(title);

	// add progress bar
	let bar = document.createElement("div");
	bar.style = "border: 1px solid #A0A0A0; min-width: 200px; max-width: 400px; width: 50%; height: 20px;margin: 0px 25px; overflow: hidden; padding: 1px 0px;";
	container.appendChild(bar);
	let inner = document.createElement("div");
	inner.style = "background: #3399FF; height: 100%; transition: width 0.5s, margin-left 0.25s;";
	inner.style.width = "50%";
	bar.appendChild(inner);

	// add footer
	let footer = document.createElement("a");
	footer.style = "position: absolute; bottom: 0px; left: 0px; right: 0px; background: black; color: white; font-size: 10px; padding: 5px 5px; text-align: right;";
	footer.textContent = "Built with ZephJS!";
	footer.setAttribute("href","http://zephjs.com");
	container.appendChild(footer);

	let loading = null;
	let loaded = null;
	let tick = 1;

	const update = ()=>{
		if (loading===null) {
			tick += 1;
			let start = (tick%150)-50;
			inner.style.width = "50%";
			inner.style.marginLeft = start+"%";
			setTimeout(update,20);
		}
		else {
			inner.style.marginLeft = "0px";
			inner.style.width = ((100/loading)*loaded)+"%";
		}
	};
	update();

	// wait for zeph:ready event
	document.addEventListener("zeph:loading",()=>{
		console.log("loading");
		loading = (loading===null ? 0 : loading) + 1;
		update();
	});
	document.addEventListener("zeph:loaded",()=>{
		console.log("loaded");
		loaded = (loaded===null ? 0 : loaded) + 1;
		update();
	});
	document.addEventListener("zeph:ready",()=>{
		console.log("ready");
		update();
		container.remove();
	});
};

let attempt = ()=>{
	if (document.body) {
		loader();
	}
	else {
		setTimeout(attempt,0);
	}
};

attempt();
